clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/UpsertGoogleImageTagsAtomicOperation.groovy

```diff
@@ -0,0 +1,99 @@
+
+import com.google.api.services.compute.model.*
+import com.netflix.spinnaker.clouddriver.data.task.Task
+import com.netflix.spinnaker.clouddriver.data.task.TaskRepository
+import com.netflix.spinnaker.clouddriver.google.config.GoogleConfigurationProperties
+import com.netflix.spinnaker.clouddriver.google.deploy.GCEUtil
+import com.netflix.spinnaker.clouddriver.google.deploy.description.UpsertGoogleImageTagsDescription
+import com.netflix.spinnaker.clouddriver.orchestration.AtomicOperation
+import org.springframework.beans.factory.annotation.Autowired
+
+/**
+ * Update the set of labels defined on an image. The newly-specified key-value pairs are added to the existing set.
+ * Values can be empty.
+ *
+ * Uses {@link https://cloud.google.com/compute/docs/reference/beta/images/setLabels}
+ */
+class UpsertGoogleImageTagsAtomicOperation implements AtomicOperation<Void> {
+  private static final String BASE_PHASE = "UPSERT_IMAGE_TAGS"
+
+  private static Task getTask() {
+    TaskRepository.threadLocalTask.get()
+  }
+
+  private final UpsertGoogleImageTagsDescription description
+
+  @Autowired
+  private GoogleConfigurationProperties googleConfigurationProperties
+
+  @Autowired
+  String clouddriverUserAgentApplicationName
+
+  UpsertGoogleImageTagsAtomicOperation(UpsertGoogleImageTagsDescription description) {
+    this.description = description
+  }
+
+  /**
+   * curl -X POST -H "Content-Type: application/json" -d '[ { "upsertImageTags": { "imageName": "my-image-123", "tags": { "some-key-1": "some-val-2" }, "credentials": "my-account-name" } } ]' localhost:7002/gce/ops
+   */
+  @Override
+  Void operate(List priorOutputs) {
+    task.updateStatus BASE_PHASE, "Initializing upsert of image tags for $description.imageName..."
+
+    def credentials = description.credentials
+    def compute = credentials.compute
+    def project = credentials.project
+    def imageName = description.imageName
+    def tags = description.tags
+    def image = GCEUtil.queryImage(project,
+                                   imageName,
+                                   credentials,
+                                   compute,
+                                   task,
+                                   BASE_PHASE,
+                                   clouddriverUserAgentApplicationName,
+                                   googleConfigurationProperties.baseImageProjects)
+
+    if (image) {
+      // Image self links are constructed like this:
+      // https://www.googleapis.com/compute/alpha/projects/rosco-oss-2/global/images/spinnaker-rosco-all-20161229193556-precise
+      def imageSelfLinkTokens = image.getSelfLink().split("/")
+      def imageProject = imageSelfLinkTokens[imageSelfLinkTokens.length - 4]
+
+      Map<String, String> originalLabels = image.getLabels()
+
+      if (originalLabels == null) {
+        originalLabels = [:]
+      }
+
+      Map<String, String> newLabels = originalLabels + tags
```

Comment:
It is with this particular operation. We could I suppose treat an empty value as indicating that a label is to be removed, but this would deviate from the behavior of the GCE api itself.

I think it's best to handle this use case if we eventually add a deleteLabels operation. For now, this is consistent with how the aws operation functions.